DataStructures/DynamicStacks/DynamicStack.java

```diff
@@ -0,0 +1,81 @@
+public class DynamicStack implements Stack{
+    //TODO: 
+    // Add two constructors: parameterless and with parameter that takes a limit
+    // If no limit is set, isFull will return false always, 
+    // otherwise it will check the current size vs the limit
+    // Add getter and setter for max nodes
+    private int numberOfNodes = 0;
+    private Integer numberOfNodesLimit = null;
+    private Node topNode = null;
+
+    public DynamicStack() {
+
+    }
+
+    public DynamicStack(Integer numberOfNodesLimit){
+        setNumberOfNodesLimit(numberOfNodesLimit);
+    }
+
+    @Override
+    public boolean push(int payload){
+        if(!isFull()){
+            Node nd = new Node(payload, topNode);
+            topNode = nd;
+            this.numberOfNodes++;
+            return true;
+        }
+        return false;        
+    } 
+
+    @Override
+    public Integer peek(){
+        if (isEmpty()){
+            return null;
+        }
+        return topNode.getPayload();
+    }
+
+    @Override
+    public Integer pop(){
+        if(isEmpty()){
+            System.out.println("No node found");
+            return null;
+        }
+        int tmpNode = topNode.getPayload();
+        topNode = topNode.getNext();
```

Comment:
On this case you have a dangling reference to the stack... Think of this, we have a stack with the following elements (the hex numbers are a symbolic representation of the memory location
```
+---+      +---+      +---+
| 3 |  ->  | 5 |  ->  | 2 |
+---+      +---+      +---+
 x03        x02        x01

topNode = x03
```
Then we call `pop()` on it. The stack will return 3 but the memory will look like this:
```
+---+      +---+      +---+
| 3 |  ->  | 5 |  ->  | 2 |
+---+      +---+      +---+
 x03        x02        x01

topNode = x02
```
The node at location `x03` will still have a reference to the rest of the stack, and that is not very clean. At this point with the current setup is not a big deal since the objects don't live for long. But on a larger application, leaving those dangling references will prevent the garbage collector from removing the nodes that you don't use anymore, and eventually the memory stack could get all filled up. 

The suggestion is to set the old topNode to point to `null` that way the node is entirely isolated:
```
+---+      +---+      +---+
| 3 |      | 5 |  ->  | 2 |
+---+      +---+      +---+
 x03        x02        x01

topNode = x02
```

Reply:
I believe I have resolved this.

Reply:
Not yet.